### ISSUE TITLE
Permanent icon and pause feature

### ISSUE THREAD

**Issue by yesdrs13** [linuxmint/timeshift#295]
**Is your feature request related to a problem? Please describe.**
Whenever I do intensive video editing, I have to go into timeshift and turn off all scheduled backups.  If the schedules are on and a backup runs during an edit, it interrupts the edit somewhat and the quality is not as great at times.  It's a little annoying to turn off the schedules all the time and, sometimes, when I'm done editing, I forget to turn them back on.

**Describe the solution you'd like**
It would be great to have a permanent icon in the app bar that you could see the status of timeshift and to have a few "pause" options you can click on that would pause for a few set length of times (1, 2, 4, 8 hours) and one option to pause indefinitely until the user is ready to resume the backup schedule in which the user will have a "un-pause" option to turn off any pause.  After the paused time has expired, timeshift would automatically turn back on the schedules again.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by MarkMoretto** [comment]
Yes.  Or, a pop-up notification when it starts running (so, not necessarily permanent).

Had an interview basically stop because my PC froze when this started running and bogged down some resources.

**Comment by yesdrs13** [comment]
In Linux Mint Cinnamon 21.3, I do get a "pop-up" when it runs and disappears when completed, but no option to "pause" it.  When you hover over the icon that pops up, it states something like "Timeshift in Progress" or "Backup in progress" or something similar to that extent.

**Comment by MarkMoretto** [comment]
Hol up.  How do I get a pop-up to appear?

I'm on MX Linux (Debian 12), if that helps.

if there is a pop-up, I agree that a Pause option would be great.  Full support of that.

**Comment by yesdrs13** [comment]
Mark, I'm sorry.  I only use Linux Mint and not sure of other distros.  I know in Linux Mint Cinnamon 20.3, there was no "pop-up", but when I upgraded to LMC 21.1 and higher, the pop-ups showed.  I think the devs for LM integrated a routine to do this starting with LM 21.  🙂  In LMC, there is an applet called "Timeshift Spy" - by nicog60, that may help rectify a pop-up issue.  You can check it out at [Timeshift Spy](https://cinnamon-spices.linuxmint.com/applets/view/327).  Hopefully, MX Linux allows applets.

**Comment by yesdrs13** [comment]
Mark, wanted to show you on my screen what pops up when Timeshift starts to backup.  When 2 cogs show up, Timeshift is processing a backup.  When it's done the backup, the cogs disappear.  This is not an applet, but directly from Timeshift.
![Screenshot from 2024-05-02 12-00-31](https://github.com/linuxmint/timeshift/assets/79882485/e66aee02-d257-41ae-9f40-b29705447e21)

**Comment by MarkMoretto** [comment]
@yesdrs13 Nice!  Thanks.

That's crispy-clean desktop!

Anyway, I was initially going to ask about which environment you're on (Cinnamon, MATE, xfce) since I'm on KDE, but I think I have to adjust my notification settings since i don't see the bell (icon on the left on the right-side of your task manager bar), so that might fix one thing.  I'll check back if not.

**Comment by yesdrs13** [comment]
Hey Mark, it's the Linux Mint Cinnamon Edition 21.3 I'm using.  With the screen...  I'm more organized on my computer than my filing cabinet!  LOL

**Comment by yesdrs13** [comment]
Mark, I was curious about MX Linux and made additional partitions and installed the XFCE (ahs) and the KDE versions.  On the XFCE, I did try out Timeshift on it and nope!  No pop up at all.  I haven't tried on KDE yet, but when I do, I'll let you know the outcome.  By the way, I've never heard of MX Linux before and thought I'd give it a shot.  It's not bad, just a few things that are "quirky", but there were some "work-arounds" I found already.

**Comment by 0x6A7232** [comment]
+1 on pause option.  Workaround as mentioned in the OP is turning off all backups that may occur within the desired pause timeframe. You may also be able to use something like the I/O rate limit feature (rsync --bwlimit flag) requested in #441 if that gets implemented (should be able to configure a rate that reduces I/O use enough to allow background backups to occur instead of causing problems).

**Comment by ygerlach** [comment]
I guess an option to "pause new backups until computer is restarted" or "pause new backups for x hours" should be quite easy to implement. I would just create a flag in the `timeshift.json` file, that blocks until a given timestamp or the `boot_id` (from `/proc/sys/kernel/random/boot_id`) was changed.

I think "pause currently running rsync backup" would also be quite interessing. I think a `SIGSTOP` and `SIGCONT` to the `AsyncTask` should be enough 🤔 
The problem might be, that the timeshift showing the gui might not be the same instance creating the backup.

**Comment by mscheifer** [comment]
I've discovered a silly workaround for this. It seems that on my system if the Timeshift GUI is open then the cron job won't run. So just opening the GUI serves to pause jobs as long as it's open. (I don't think it would work if the backup has already started though). This might not be the same on every system. I'm on Mint 22.2 for reference.

**Just opening the Timeshift GUI** is potentially a temporary workaround.

**Comment by 0x6A7232** [comment]
I see a suggestion that could be added as a toggle (maybe on by default so newbies don't get hit by it, but highly visible in case power users don't want it) to add / use ionice to the `timeshift-*` rsync cron job: `ionice -c 3` ... <cron job> -- this makes the backup job run as an idle task, which, honestly, it should be by most people's defaults, right? Unless it's an ondemand task, but even then it would be nice to have the ability to launch it as idle priority.

Thoughts?

**Comment by ygerlach** [comment]
>  use ionice

inside timeshift are still code snippets for using ionice (example: `RsyncTask.vala`). I dont know why it [was removed](https://github.com/linuxmint/timeshift/commit/198068461cff61c4e6c3f75d099864fe097a9903) 9 years ago. Maybe there was a good reason?

But if ionice is available (we should check that, a snapshot does not fail because of a missing ionice binary)  i think we can use it again for scripted snapshots.

Maybe we can just do the required syscalls ourselfs and dont need to depend on external binarys for such a simple problem. But i dont know how ionicing works, so i have to investigate.

EDIT:
Its just a simple syscall.

ionice only calls `syscall(SYS_ioprio_set, ...)`

More details can be found in `man 2 ioprio_set`. I think instead of using the shell command and relying on it, its better to use the syscall directly.

**Comment by teejee2008** [comment]
There reason Timeshift doesn't use ionice and does not have a pause option is because Rsync snapshots need to complete as soon as possible, to avoid capturing a system state that is inconsistent.
If there is a lot of disk/system activity going on and the snapshot takes a long time to complete, then restoring that snapshot later will cause problems on the restored system.
Ideally the snapshot should be atomic, but since that is not possible with rsync snapshots, the snapshots should be allowed to complete as quickly as possible (without pausing it or using ionice).

**Comment by 0x6A7232** [comment]
> There reason Timeshift doesn't use ionice and does not have a pause option is because Rsync snapshots need to complete as soon as possible, to avoid capturing a system state that is inconsistent. If there is a lot of disk/system activity going on and the snapshot takes a long time to complete, then restoring that snapshot later will cause problems on the restored system. Ideally the snapshot should be atomic, but since that is not possible with rsync snapshots, the snapshots should be allowed to complete as quickly as possible (without pausing it or using ionice).

So, on my system, I am backing up my root + home filesystem which is a ~400GB partition (50% utilized) on a 4TB Samsung 870 EVO SATA drive, to a ~500GB (52% utillzed) partition on a 10TB 7200RPM WD Red Pro NAS drive.  It takes about 5-10 minutes.  I'm on the Liquorix kernal 6.,16.12-1 so I'm using the bfq scheduler for both.  More on that at the end if you're interested.

During that time, every process that is significantly bottlenecked by disk I/O (and this includes all Steam games, at least the ones using Proton, I think this is a bug with Steam / Proton because they shouldn't be using the disk heavily at all, and I can actually see via the activity LED that they normally aren't during non-backup operation, but I don't know how to check what is causing this) will experience complete freezes.  Usually I can still hear background music playing.  I can alt-tab and browse the Internet.  After 10-45 seconds, it will unfreeze and resume as if no time at all has passed.

The only way around this so far is to either launch the Timeshift window to pause backups (I have hourly backups) or to disable the scheduled backups until I'm done gaming.

I hear your arguments about preserving the system state as closely as possible.  It's why I have hourly backups -- if a backup is in the middle of a system change I can just choose another hour, no big deal.

However, shouldn't this be left up to the user?  Yours seems like a solid argument for having the DEFAULTS set to normal priority, but not to remove the option for background backup jobs altogether. 

This also seems like a solid argument for having a brief description of the pros and cons of setting background priority by the option checkmark to set backup tasks as such.

Isn't there also an option between background priority and normal priority?

PS re: bfq
According to the AI I was trying to troubleshoot this issue with,

> Liquorix deliberately sets BFQ as the default for all block devices because:
> - Modern BFQ (especially the version in 5.15+ and newer kernels) detects SSDs automatically and switches to an almost “none”-like behavior on them (very low overhead).
> - It still gives slightly better interactive feel under mixed loads (e.g., compiling + gaming + backup) even on SSDs.
> - It simplifies the rules: one excellent low-latency scheduler for everything instead of per-device heuristics.

**Comment by mscheifer** [comment]
> There reason Timeshift doesn't use ionice and does not have a pause option is because Rsync snapshots need to complete as soon as possible, to avoid capturing a system state that is inconsistent. If there is a lot of disk/system activity going on and the snapshot takes a long time to complete, then restoring that snapshot later will cause problems on the restored system. Ideally the snapshot should be atomic, but since that is not possible with rsync snapshots, the snapshots should be allowed to complete as quickly as possible (without pausing it or using ionice).

That makes sense, but I think the original idea here (for this issue) was to pause the cronjob that starts the snapshot in the first place. That wouldn't create any inconsistent snapshots.

**Comment by 0x6A7232** [comment]
Yeah, good point on the original idea.
As for ionice, I only brought it up because it seems like a decent solution... like... since when do you _ever_ want a **_scheduled_** backup job to I?O jam your system when it is _in use_?
Never.  The answer is never.  When the system is in use, the user has (at least marginally) higher priority than scheduled tasks, forever and always.  See also: Windows Updates. 
"Well if you want that, you just set up the cron job via the Terminal yourself!"
See also: Windows Updates (you can use the group policy editor and/or regedit to get Windows to stop forcing system restarts and mandatory updates, but we all know the right solution is to have a toggle in the Windows Update setting.  Microsoft is just trying to nanny-sit their users because they have a habit of making choices MS doesn't agree with).
Now, of course this is your project and you can do what you want with it.  I'm just giving you my viewpoint.  Feel free to disregard.  Sorry for the slight rant, I don't mean to be offensive.

Oh.  Actually, that brings up another potential solution:
You could have an option you can set for certain backups to only run when the system is at an idle state for a certain period of time.  Not sure how difficult that would be, but it would avoid the problem in another way.

**Comment by ygerlach** [comment]
It is impossible to avoid data corruption using rsync. Especially if things like sqlite files are involved (browser bookmarks/cookies anyone?). Sure one could improve his odds by making rsync run faster, but in the end the possibility of corruption might be slim but never zero.

So instead of that, i think we should just embrace the fact that rsync is not perfect, if the system is running and give it slightly lower io prio so it does not stall user activity. This will likely add < 1% of runtime to the average idle rsync snapshot. So virtually no increased risk here.
Additionally we should add a fair warning to the mode selection, that rsync is known to produce corrupt snapshots under bad circumstances.

> system is at an idle state for a certain period of time

i fear, that some systems never reach this level of idle and would therefore never create a snapshot. But i like the idea. Maybe with a timeout, when it starts taking a snapshot anyways?


> but I think the original idea here (for this issue) was to pause the cronjob that starts the snapshot in the first place.

that was, what my idea here was for:
https://github.com/linuxmint/timeshift/issues/295#issuecomment-3289794044

I guess i will implement something like that, but i will probably take a month or two.

**Comment by teejee2008** [comment]
> So instead of that, i think we should just embrace the fact that rsync is not perfect, if the system is running and give it slightly lower io prio so it does not stall user activity. This will likely add < 1% of runtime to the average idle rsync snapshot. So virtually no increased risk here. Additionally we should add a fair warning to the mode selection, that rsync is known to produce corrupt snapshots under bad circumstances.

Yes, the risk is low. Maybe its not worth worrying about.

> i fear, that some systems never reach this level of idle and would therefore never create a snapshot. But i like the idea. Maybe with a timeout, when it starts taking a snapshot anyways?

I agree. Defining the criteria for idleness will not be easy. If we are checking for idleness by checking disk I/O activity, then it will never be idle, unless the user is away from the system. One thing that is useful to check is whether apt/dpkg is running before starting the snapshot. If apt/dpkg is running TS should wait for that to complete. But then again, it is rare for background updates to be running at the same time that the snapshot is being created.

> that was, what my idea here was for: [#295 (comment)](https://github.com/linuxmint/timeshift/issues/295#issuecomment-3289794044)
> 
> I guess i will implement something like that, but i will probably take a month or two.

Pausing the schedule for a fixed duration (like 2, 4, or 8 hours) is indeed useful.

Pausing a running snapshot on the other hand is not recommended. It is better to cancel the snapshot than to pause it.

**Comment by 0x6A7232** [comment]
+1 to the idea of not running rsync if apt / dpkg / dnf / snap / pacman / portage / any other package manager is running.  I don't know how to check for all of those, maybe something like the AI snippets at the end, but perhaps the same hook as is currently used for not running snapshots when Timeshift GUI is active could be used as the entry point for this check if it returns positive?


<details><summary>AI slop</summary>
Check package manager

```
#!/bin/bash

package_managers=("apt" "yum" "dnf" "zypper" "pacman" "snap" "flatpak")
echo "Installed package managers on this Linux system:"
for manager in "${package_managers[@]}"; do
    if command -v "$manager" > /dev/null 2>&1; then
        echo "- $manager"
    fi
done
```

Check if manager is currently running

```
#!/bin/bash

package_managers=("apt" "yum" "dnf" "zypper" "pacman" "snap" "flatpak")
echo "Checking for running package managers:"

for manager in "${package_managers[@]}"; do
    if pgrep -x "$manager" > /dev/null; then
        echo "- $manager is currently running"
    fi
done
```

</details>

Apologies for AI slop but I'm sure the same mental coder circuits tearing apart the snippets are also automatically finding the correct way to accomplish this (or maybe the AI got it right ha, who knows, it was just a search AI).

**Comment by ygerlach** [comment]
> Check if manager is currently running

if we do that, we have to keep in mind, that some users have a hook in their package manager to run timeshift before / after updates. If timeshift then waits for the package manager to terminate, it will probably hang for ever. Also stopping the user from downgrading timeshift or uninstalling the hook...

I guess that are two features:
1. Allowing users to pause the cron job for some time
2. Detecting package managers running and wait for them (cron snapshot), or warn the user (manual snapshot)

To detect if timeshift was triggered by a package manager it could walk its parent processes, similar to what `get_user_process` does and check if one of the parents is apt, pacman, dnf, ...

**Comment by 0x6A7232** [comment]
EDIT: looked up how to use collapsible text when using markdown to avoid irritation with AI content (sorry ygerlach).

> > Check if manager is currently running
> 
> if we do that, we have to keep in mind, that some users have a hook in their package manager to run timeshift before / after updates. If timeshift then waits for the package manager to terminate, it will probably hang for ever. Also stopping the user from downgrading timeshift or uninstalling the hook...
> 
> I guess that are two features:
> 
>     1. Allowing users to pause the cron job for some time
> 
>     2. Detecting package managers running and wait for them (cron snapshot), or warn the user (manual snapshot)
> 
> 
> To detect if timeshift was triggered by a package manager it could walk its parent processes, similar to what `get_user_process` does and check if one of the parents is apt, pacman, dnf, ...

Maybe add a timeout for the wait if check finds package manager running?

Also, wouldn't it be easier to just check if Timeshift was started by the user somehow?  

Hmph.  After consulting an AI, apparently detecting who actually launched a process isn't trivial and checking a list of package manager processes / lock files using `pgrep` or similar with a timeout included to avoid loops would be easier..

tl;dr: Something so simple seems to be a PITA to work around, and probably different for every distro, yikes.  

<details><summary>AI maybe-slop:</summary>

> The thread discusses a feature request for Timeshift to add a system tray icon for status monitoring and a way to temporarily pause scheduled backups (e.g., for 1-4 hours or indefinitely) without disabling them entirely, primarily to avoid disruptions during I/O-intensive user tasks like video editing. Backups can cause system slowdowns or freezes due to high disk activity, and users often forget to re-enable schedules after manual disables. Related pain points include backups conflicting with package managers (leading to potential hangs or inconsistent snapshots) and the need for better background execution without interrupting foreground work.
> 
> The specific reply you highlighted appears to propose or discuss using process checks (e.g., via pgrep) to detect active package managers like apt, dnf, or pacman before starting a backup, waiting if they're running to avoid conflicts. However, it raises concerns about incomplete detection (missing less common managers) and potential infinite loops in scenarios where Timeshift is hooked to run before/after updates—e.g., if the backup waits for the package manager to finish, but the package manager is paused waiting for the backup hook to complete, creating a deadlock.
> 
> Regarding your suggestion: Checking if Timeshift was launched interactively by the user (e.g., via a desktop environment or launcher) could be a viable alternative to exhaustively detecting every package manager, but it wouldn't necessarily be easier or more reliable in all cases. Here's why:
> 
> - **Pros of your approach:** It avoids maintaining an ever-growing list of package managers (apt, dnf, pacman, yum, zypper, emerge, etc.) and their variants/process names/lock files. Detection could be simple, like checking if the `$DISPLAY` environment variable is set (indicating a graphical session) or if stdin is a TTY (`isatty(0)` in code). This approximates "user-launched" vs. automated (cron/hook), allowing different behaviors—e.g., only perform waits/checks in interactive mode, or skip them in non-interactive to prevent loops.
> 
> - **Cons and challenges:** 
> 
>   - Desktop environments vary (GNOME, KDE, Cinnamon, etc.), and parent processes might be generic (e.g., `bash`, `sh`, or the DE's shell). Spotlight-like launchers (e.g., rofi, dmenu, Albert, or GNOME's search) typically inherit the DE's environment (including `$DISPLAY`), so they'd count as "user-launched," which is fine for your intent.
>   - Edge cases: Headless servers or SSH sessions might lack `$DISPLAY` but still be user-initiated CLI runs. Automated scripts could be run in a graphical session (inheriting `$DISPLAY`). Cron jobs and package hooks usually lack both `$DISPLAY` and TTY, but custom user setups could blur this.
>   - It doesn't directly address the root conflict (backups running concurrently with package ops, risking inconsistencies like partial file copies). You'd still need some form of conflict avoidance, just gated by launch detection.
> 
> A more targeted tweak to the reply's package manager detection idea could achieve similar loop avoidance without broad launch checks: In code, get the parent process ID (via `getppid()`), then inspect its command name (e.g., `ps -p $ppid -o comm=`). If the parent matches a short list of known package managers (e.g., `apt`, `dpkg`, `pacman`), assume it's a hook invocation and skip the wait/check entirely. This prevents deadlocks in hooked scenarios while allowing normal waits otherwise (e.g., for cron backups coinciding with manual updates). Maintaining a list of ~10 common manager commands is simpler than full process/lock scanning across the system, and it doesn't require DE/launcher-specific logic.
> 
> ### Best Solution for the Issue(s)
> The thread's core problems—disruptive backups, forgetful re-enabling, package conflicts/loops, and snapshot consistency—call for a multi-faceted fix prioritizing usability, safety, and minimal code complexity. Based on the discussions (including I/O prioritization, pause mechanics, and detection trade-offs), here's what I see as the optimal approach, building on the developer's stated plans:
> 
> **1. Core Pause Feature (Addresses Main User Pain Point):**
> 
> - Add a system tray icon showing backup status (idle, running, paused) with right-click options to pause schedules for preset durations (1/2/4/8 hours) or indefinitely.
> - Implement via a flag in the config file (`timeshift.json` or similar) with a timestamp or boot ID. When paused, cron jobs check the flag and skip execution until the pause expires or is manually cleared.
> - Automatic resumption after pause, with notifications. This is safer than suspending running rsync processes (which risks corruption via `SIGSTOP`/`SIGCONT`); instead, cancel if needed and restart later.
> - Tie into existing behavior where an open GUI locks concurrent runs, treating it as an "implicit pause."
> 
> **2. I/O Throttling for Non-Disruptive Background Runs (Prevents Interruptions Without Pausing):**
> 
> - Always run backups at idle I/O priority using `ionice -c 3` (or equivalent syscall `ioprio_set`). This lets foreground tasks (e.g., video editing) take precedence, making backups "invisible" to users.
> - Historical code in Timeshift supported this but was removed; re-adding it is low-effort and compares favorably to systems like Windows Update. Drawback: Slightly longer backup times, but the consistency risk is low for quick rsync snapshots.
> - Optional: Add UI warnings about potential (rare) corruption if snapshots occur during heavy writes (e.g., to databases).
> 
> 
> **3. Package Manager Conflict Handling with Loop Prevention (Builds on the Specific Reply):**
> 
> - Before starting a backup, check for active managers via `pgrep` on a curated list of processes (apt-get, dpkg, pacman, etc.) and/or lock files (e.g., `/var/lib/dpkg/lock`).
> - If detected, wait with a hard timeout (e.g., 5-10 minutes) before proceeding or aborting safely. This avoids infinite loops without skipping checks entirely.
> - Add the parent process check (as I described above) to detect hook contexts and bypass the wait if the parent is a package manager. This directly resolves deadlock risks in update hooks.
> - Make it configurable (e.g., via flags like `--no-pkg-check` for hooks/scripts), and log skips for debugging.
> 
> **4. Idle Detection as a Fallback (Enhances Reliability):**
> 
> - Delay automated backups until the system is idle (e.g., low disk I/O or CPU for 5+ minutes, using tools like `iostat` or simple checks). If no idle window within a threshold (e.g., 1 hour), force execution with ionice.
> - This complements package checks but avoids complex "idle" definitions (e.g., user away via screensaver hooks).
> 
> 
> This combination aligns with the thread's consensus: Focus on schedule pausing over process suspension, prioritize low-disruption execution, and handle conflicts without exhaustive detection. It's feasible for the Timeshift codebase (GTK-based GUI, rsync backend) and could be prototyped via code execution tools if needed (e.g., using Python's `psutil` for process checks). Development timeline mentioned in the thread was 1-2 months, so by late 2025, check for updates in the repo. 

The prompt:

> Have a look at this thread about Timeshift: https://github.com/linuxmint/timeshift/issues/295 
> Specifically, focus on this response: https://github.com/linuxmint/timeshift/issues/295#issuecomment-3678656438 
> Would it be easier to just check if Timeshift was launched by the user (check if launched by the Desktop Environment?  But what about Spotlight-esque launchers?) rather than attempting to detect every package manager to avoid the loop?
> Let me know if you can access the thread and the specific reply I'm referring to.
> What would be the best solution for the issue(s) mentioned in the thread?
</details>


EDIT:

<details><summary>Probably garbage but here's what the AI would code to implement all of that:</summary>

Prompt:

> https://github.com/linuxmint/timeshift/blob/master/src/Gtk/MainWindow.vala
> https://github.com/linuxmint/timeshift/blob/master/src/Core/Main.vala
> Output the code you would use to make each of the changes you proposed earlier.  Use ``` (triple-tick) to encapsulate code blocks, followed by the change in diff format, also within triple-tick code blocks, as well as a human readable description of the changes made and reasons for them.

Response (thought for 42s):

### 1. Core Pause Feature

This change adds a system tray icon to the GUI for displaying backup status and providing pause options (e.g., 1/2/4/8 hours or indefinitely). It modifies `MainWindow.vala` to include the tray icon and menu items for pausing. In `Main.vala`, it adds logic to check a pause flag in the config file (`timeshift.json`) before running scheduled backups, using a timestamp to track pause duration. This prevents backups during user-specified periods without disabling schedules entirely, addressing forgetfulness in re-enabling. The pause is stored as a JSON key like `"pause_until": "timestamp"` or `"pause_duration": "hours"`. Notifications are sent on resumption. This is low-risk as it doesn't alter running processes.

vala
```
// Add to MainWindow.vala, after initializing toolbar

private Gtk.StatusIcon tray_icon;
private Gtk.Menu tray_menu;

private void init_tray_icon() {
    tray_icon = new Gtk.StatusIcon();
    tray_icon.set_from_icon_name("timeshift");
    tray_icon.set_tooltip_text(_("Timeshift - System Snapshots"));
    tray_icon.set_visible(true);
    tray_icon.activate.connect(on_tray_activate);

    tray_menu = new Gtk.Menu();

    var item_pause_1h = new Gtk.MenuItem.with_label(_("Pause for 1 hour"));
    item_pause_1h.activate.connect(() => pause_backups(1));
    tray_menu.append(item_pause_1h);

    var item_pause_2h = new Gtk.MenuItem.with_label(_("Pause for 2 hours"));
    item_pause_2h.activate.connect(() => pause_backups(2));
    tray_menu.append(item_pause_2h);

    var item_pause_4h = new Gtk.MenuItem.with_label(_("Pause for 4 hours"));
    item_pause_4h.activate.connect(() => pause_backups(4));
    tray_menu.append(item_pause_4h);

    var item_pause_8h = new Gtk.MenuItem.with_label(_("Pause for 8 hours"));
    item_pause_8h.activate.connect(() => pause_backups(8));
    tray_menu.append(item_pause_8h);

    var item_pause_indefinite = new Gtk.MenuItem.with_label(_("Pause Indefinitely"));
    item_pause_indefinite.activate.connect(() => pause_backups(-1));
    tray_menu.append(item_pause_indefinite);

    var item_resume = new Gtk.MenuItem.with_label(_("Resume Backups"));
    item_resume.activate.connect(resume_backups);
    tray_menu.append(item_resume);

    tray_menu.show_all();
}

private void on_tray_activate() {
    tray_menu.popup(null, null, null, 3, Gtk.get_current_event_time());
}

private void pause_backups(int hours) {
    // Update config with pause timestamp
    DateTime now = new DateTime.now_local();
    DateTime pause_until = (hours > 0) ? now.add_hours(hours) : now.add_years(100); // Indefinite as far future
    json_set_string(config, "pause_until", pause_until.format("%Y-%m-%d %H:%M:%S"));
    save_app_config();
    tray_icon.set_tooltip_text(_("Backups Paused"));
    gtk_do_events();
    show_notification(_("Backups Paused"), _("Scheduled backups are paused until ") + pause_until.format("%H:%M"));
}

private void resume_backups() {
    json_remove_key(config, "pause_until");
    save_app_config();
    tray_icon.set_tooltip_text(_("Timeshift - Idle"));
    show_notification(_("Backups Resumed"), _("Scheduled backups have resumed."));
}

private void show_notification(string title, string message) {
    var notification = new Notify.Notification(title, message, "timeshift");
    notification.show();
}

// Call init_tray_icon() in MainWindow constructor after init_ui_statusbar();
init_tray_icon();

// Add to Main.vala, in backup execution logic (e.g., before create_snapshot in scheduled runs)

private bool is_paused() {
    string pause_str = json_get_string(config, "pause_until", "");
    if (pause_str.length > 0) {
        DateTime pause_until = new DateTime.from_string(pause_str, date_format);
        DateTime now = new DateTime.now_local();
        if (now.compare(pause_until) < 0) {
            return true;
        } else {
            json_remove_key(config, "pause_until");
            save_app_config();
        }
    }
    return false;
}

// In scheduled backup check (e.g., in cron or run logic):
if (is_paused()) {
    log_debug(_("Backup paused by user"));
    return;
}
```
diff
```
diff --git a/src/Gtk/MainWindow.vala b/src/Gtk/MainWindow.vala
--- a/src/Gtk/MainWindow.vala
+++ b/src/Gtk/MainWindow.vala
@@ -80,6 +80,8 @@ class MainWindow : Gtk.Window{
         init_ui_snapshot_list();
 
         init_ui_statusbar();
+		
+		init_tray_icon();  // Added for system tray pause functionality
 
         if (App.live_system()){
             btn_backup.sensitive = false;
@@ -300,6 +302,52 @@ class MainWindow : Gtk.Window{
         return true;
     }
 
+	// Added tray icon and pause menu
+	private Gtk.StatusIcon tray_icon;
+	private Gtk.Menu tray_menu;
+
+	private void init_tray_icon() {
+	    tray_icon = new Gtk.StatusIcon();
+	    tray_icon.set_from_icon_name("timeshift");
+	    tray_icon.set_tooltip_text(_("Timeshift - System Snapshots"));
+	    tray_icon.set_visible(true);
+	    tray_icon.activate.connect(on_tray_activate);
+
+	    tray_menu = new Gtk.Menu();
+
+	    var item_pause_1h = new Gtk.MenuItem.with_label(_("Pause for 1 hour"));
+	    item_pause_1h.activate.connect(() => pause_backups(1));
+	    tray_menu.append(item_pause_1h);
+
+	    // ... (similar for 2h, 4h, 8h, indefinite, resume)
+
+	    tray_menu.show_all();
+	}
+
+	private void on_tray_activate() {
+	    tray_menu.popup(null, null, null, 3, Gtk.get_current_event_time());
+	}
+
+	private void pause_backups(int hours) {
+	    // ... (as in code block)
+	}
+
+	private void resume_backups() {
+	    // ... (as in code block)
+	}
+
+	private void show_notification(string title, string message) {
+	    // ... (as in code block)
+	}
+
+	// End of additions for pause feature
+
     private Gtk.MenuItem create_menu_item(
         string label_text, string icon_name_stock, string icon_name_custom,
         int icon_size, string tooltip_text = ""){
diff --git a/src/Core/Main.vala b/src/Core/Main.vala
--- a/src/Core/Main.vala
+++ b/src/Core/Main.vala
@@ -500,6 +500,25 @@ public class Main : GLib.Object{
         // ... existing code ...
     }
 
+	// Added pause check
+	private bool is_paused() {
+	    string pause_str = json_get_string(config, "pause_until", "");
+	    if (pause_str.length > 0) {
+	        DateTime pause_until = new DateTime.from_string(pause_str, date_format);
+	        DateTime now = new DateTime.now_local();
+	        if (now.compare(pause_until) < 0) {
+	            return true;
+	        } else {
+	            json_remove_key(config, "pause_until");
+	            save_app_config();
+	        }
+	    }
+	    return false;
+	}
+
+	// In backup method, add:
+	if (is_paused()) { return; }
+
     public void create_snapshot(){
         // ... existing backup logic ...
     }
```

### 2. I/O Throttling for Non-Disruptive Background Runs

This change ensures backups run at idle I/O priority to avoid disrupting foreground tasks like video editing. It modifies the RsyncTask execution in `Main.vala` to prefix rsync commands with `ionice -c 3` (idle class). This is applied to all backup operations, making them less intrusive while slightly increasing duration. Historical code supported this but was removed; re-adding is straightforward and aligns with systems like Windows Update. No GUI changes needed, as it's backend-only.

vala
```
// Add to Main.vala, in RsyncTask or backup execution (assuming task.execute() calls rsync)

public class RsyncTask : AsyncTask {
    // Override or modify execute_command to include ionice
    protected override void execute_command(string cmd) {
        string throttled_cmd = "ionice -c 3 " + cmd;  // Run at idle I/O priority
        base.execute_command(throttled_cmd);
    }
}

// Alternatively, if no RsyncTask class, in Main.create_snapshot():
string rsync_cmd = "ionice -c 3 rsync -a --exclude=...";  // Prefix rsync calls
task.execute(rsync_cmd);
```
diff
```
diff --git a/src/Core/Main.vala b/src/Core/Main.vala
--- a/src/Core/Main.vala
+++ b/src/Core/Main.vala
@@ -900,7 +900,11 @@ public class Main : GLib.Object{
 
     public void create_snapshot(){
         // ... existing prep
-		task.execute();  // Assuming this runs rsync
+		
+		// Added I/O throttling
+		task.prg_prefix = "ionice -c 3";  // Or modify task.cmd to include "ionice -c 3 "
+		task.execute();
+		
         // ... existing post
     }
 
@@ -1000,6 +1004,11 @@ public class RsyncTask : AsyncTask {  // If separate class exists
     protected override void execute() {
-		rsync_command();
+		// Added throttling
+		string cmd = "ionice -c 3 " + get_rsync_cmd();
+		execute_command(cmd);
     }
+	
+	private string get_rsync_cmd() {
+	    // Existing rsync command building
+	}
 }
```

### 3. Package Manager Conflict Handling with Loop Prevention

This adds checks before backups in `Main.vala` to detect running package managers (e.g., via `pgrep apt|dpkg|pacman`) and wait up to 10 minutes. It includes a parent process check using `getppid()` and `ps` to skip if invoked by a package hook, preventing deadlocks. A curated list of managers ensures coverage without exhaustiveness. Configurable via `--no-pkg-check`. This builds on the thread's proposal, enhancing safety for snapshot consistency.

vala
```
// Add to Main.vala, before backup start in create_snapshot()

private bool check_package_managers() {
    string[] managers = {"apt", "apt-get", "dpkg", "pacman", "dnf", "yum", "zypper"};
    foreach (string mgr in managers) {
        string stdout, stderr;
        exec_sync("pgrep -x " + mgr, out stdout, out stderr);
        if (stdout.strip().length > 0) {
            return true;  // Active
        }
    }
    return false;
}

private bool is_hook_invocation() {
    Pid ppid = Posix.getppid();
    string cmd = "ps -p %d -o comm=".printf(ppid);
    string stdout, stderr;
    exec_sync(cmd, out stdout, out stderr);
    string parent = stdout.strip();
    return (parent in new string[]{"apt", "dpkg", "pacman"});  // Check against list
}

private void wait_for_package_managers(int timeout_min = 10) {
    if (is_hook_invocation()) {
        log_debug(_("Hook detected - skipping package check"));
        return;
    }
    int waited = 0;
    while (check_package_managers() && waited < timeout_min * 60) {
        Posix.sleep(60);  // Wait 1 min
        waited += 60;
    }
    if (waited >= timeout_min * 60) {
        log_error(_("Timeout waiting for package managers"));
        // Abort or proceed
    }
}
```
diff
```
diff --git a/src/Core/Main.vala b/src/Core/Main.vala
--- a/src/Core/Main.vala
+++ b/src/Core/Main.vala
@@ -800,6 +800,9 @@ public class Main : GLib.Object{
 
     public void create_snapshot(){
+		// Added package conflict check
+		wait_for_package_managers();
+		
         // ... existing backup logic ...
     }
 
@@ -1200,6 +1203,30 @@ public class Main : GLib.Object{
+	// Added helper functions
+	private bool check_package_managers() {
+	    // ... (as in code block)
+	}
+
+	private bool is_hook_invocation() {
+	    // ... (as in code block)
+	}
+
+	private void wait_for_package_managers(int timeout_min = 10) {
+	    // ... (as in code block)
+	}
 }
```

### 4. Idle Detection as a Fallback

This delays automated backups until the system is idle (low disk I/O via `iostat` check for 5+ minutes). If no idle window in 1 hour, force with ionice. Added to `Main.vala` for scheduled runs. This complements other features, using simple thresholds to avoid complexity. Requires `iostat` (from sysstat package).

vala
```
// Add to Main.vala, for scheduled backups

private bool is_system_idle(int threshold_sec = 300) {  // 5 min
    string cmd = "iostat -d 1 2 | tail -n +4 | awk '{print $4}'";  // Get tps (transactions/sec)
    string stdout, stderr;
    exec_sync(cmd, out stdout, out stderr);
    double tps = double.parse(stdout.strip());
    return (tps < 10.0);  // Arbitrary low threshold
}

private void wait_for_idle(int max_wait_min = 60) {
    int waited = 0;
    while (!is_system_idle() && waited < max_wait_min * 60) {
        Posix.sleep(60);
        waited += 60;
    }
    if (waited >= max_wait_min * 60) {
        log_debug(_("Forcing backup after timeout"));
    }
}
```
diff
```
diff --git a/src/Core/Main.vala b/src/Core/Main.vala
--- a/src/Core/Main.vala
+++ b/src/Core/Main.vala
@@ -800,6 +800,9 @@ public class Main : GLib.Object{
 
     public void create_snapshot(){  // For scheduled
+		// Added idle detection
+		wait_for_idle();
+		
         // ... existing
     }
 
@@ -1300,6 +1303,20 @@ public class Main : GLib.Object{
+	// Added idle helpers
+	private bool is_system_idle(int threshold_sec = 300) {
+	    // ... (as in code block)
+	}
+
+	private void wait_for_idle(int max_wait_min = 60) {
+	    // ... (as in code block)
+	}
 }
```
</details>

**Comment by ygerlach** [comment]
@0x6A7232 please stop dumping ai garbage onto us. [That is rude.](https://distantprovince.by/posts/its-rude-to-show-ai-output-to-people/)

I proposed an option to detect if timeshift was invoked by a specific program (package manager or cron):
Just iterate its parent processes. We have functions for that already. It probably additional ~10 lines of vala code.

I also proposed an option to pause cron jobs for sometime or until next boot.

**Comment by 0x6A7232** [comment]
> [@0x6A7232](https://github.com/0x6A7232) please stop dumping ai garbage onto us. [That is rude.](https://distantprovince.by/posts/its-rude-to-show-ai-output-to-people/)
> 
> I proposed an option to detect if timeshift was invoked by a specific program (package manager or cron): Just iterate its parent processes. We have functions for that already. It probably additional ~10 lines of vala code.
> 
> I also proposed an option to pause cron jobs for sometime or until next boot.

Sorry about that.  I figured if I preempted every AI generated output with "AI slop ahead", anyone not morbidly curious about it could just roll past it, my bad.  I did realize that it could be annoying for many people, hence the trigger warning.  

Ok, I looked up how to put it in collapsible text, now it won't bother anyone not interested in it, hopefully that's much better.

Also, cool on the proposals.  I see the one for setting priority since you linked it, but not the one for pausing?  Maybe I missed it.

**Comment by ygerlach** [comment]
>  I see the one for setting priority since you linked it, but not the one for pausing?

The one for pausing:
https://github.com/linuxmint/timeshift/issues/295#issuecomment-3289794044

I already started to implement that (but i haven't created a PR yet). See: https://github.com/ygerlach/timeshift/tree/pause_snapshots

My biggest problem is the question, how to make the ui. I think for now, i would do a drop down with options like:
- Pause 4h
- Pause 8h
- Pause 12h
- Pause until shutdown

**Comment by 0x6A7232** [comment]
> > I see the one for setting priority since you linked it, but not the one for pausing?
> 
> The one for pausing: [#295 (comment)](https://github.com/linuxmint/timeshift/issues/295#issuecomment-3289794044)
> 
> I already started to implement that (but i haven't created a PR yet). See: https://github.com/ygerlach/timeshift/tree/pause_snapshots
> 
> My biggest problem is the question, how to make the ui. I think for now, i would do a drop down with options like:
> 
>     * Pause 4h
> 
>     * Pause 8h
> 
>     * Pause 12h
> 
>     * Pause until shutdown

Maybe add half an hour?  Good for quick tasks.  Alternatively, leave it as-is and have an option to immediately resume once in a paused state.  So for the half hour you just put 4h, do your thing, pop back in and un-pause it.

ty for links!
